src/pluto/components/outlier_report.py

```diff
@@ -0,0 +1,89 @@
+
+
+class OutlierReport:
+    def __init__(self, data, v1, v2 ,condo, mapped):
+        self.df = data
+        self.v1 = v1
+        self.v2 = v2
+        self.condo = condo
+        self.mapped = mapped
+
+    def __call__(self):
+        st.header("Outlier Analysis")
+
+        if self.v1 not in self.versions:
+            st.info("There is no outlier report available for selected version.")
+            return           
+
+        self.display_dataframe("building_area_increase")
+
+        self.display_dataframe("unitsres_resarea")
+
+        self.display_dataframe("lotarea_numfloor")
+
+    def display_dataframe(self, field):
+        df = self.fetch_dataframe(field)
+        
+        if df.empty:
+            st.markdown(self.markdown_dict[field])
+            st.write("There are no outliers for this check.")
+            st.info(self.info_dict[field])
+        else:
+            st.markdown(self.markdown_dict[field])
+
+            AgGrid(df)
+
+            st.write(f"There are {df.shape[0]} outliers in total.")
+            st.info(self.info_dict[field])
+
+    def fetch_dataframe(self, field):
+        records = [i["values"] for i in self.v1_outlier_records if i["field"] == field][0]
+
+        if records:
+            df = pd.DataFrame(records)
+
+            if field == "building_area_increase":
+                df = df.drop(columns=["pair"])
+
+            df["bbl"] = pd.to_numeric(df["bbl"], downcast="integer")
+            return df
+        else:
+            return pd.DataFrame()
+
+    @property
+    def versions(self):
+        return self.df.v.unique()
+    
+    @property
+    def outlier_records(self):
+        return self.df.loc[
+            (self.df.condo == self.condo) & (self.df.mapped == self.mapped) & (self.df.v == self.v1),
+            :,
+        ].to_dict("records")
+
+    @property
+    def v1_outlier_records(self):
+        return [i["outlier"] for i in self.outlier_records if i["v"] == self.v1][0]
+
+    @property
+    def version_pair(self):
+        return f"{self.v1}-{self.v2}"
+
+    @property
+    def markdown_dict(self):
+        return {
+            "building_area_increase": f"### Table of BBLs with Unreasonable Increase in Building Area {self.version_pair}",
+            "unitsres_resarea": f"### Report of BBLs with buildings containing unreasonably small apartments",
+            "lotarea_numfloor": f"### Table of BBLs where bldgarea/lotarea > numfloors*2",
+        }
+    
+    @property
+    def info_dict(self):
+        return {
+            "building_area_increase": "The table displays all BBLs where building area is more than doubled since previous version.",
+            "unitsres_resarea": "The table displays all BBLs where unitsres is more than 50 and resarea is greater than 0 but the ratio of resarea:unitsres is less than 300.",
+            "lotarea_numfloor": "The table displays all BBLs where the ratio of bldgarea:lotarea is more than twice numfloors.",
+        }
```

Comment:
No new line at end of file makes me worry that this wasn't formatted with black?